hash_practice/exercises.py

```diff
@@ -2,18 +2,39 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n log (n))
+        Space Complexity: O (n)
     """
-    pass
+    anagrams = {}
+    for word in strings:
+        alphabet = [0] * 26

```

Comment:
OK, this is an interesting choice of key for your hash table, that will use a little extra space but may have other benefits in that the amount of space for each individual key is the same. (Interesting is not wrong! This is a perfectly fine solution at this level, I just wanted to talk about your choice some.)

Usually, solutions to this problem will use a sorted version of the characters in the word. In that case, each key will basically be a string the length of the words with those letters. As most English words have less than 26 letters, this will generally be a fewer number of characters than 26.

Further, your solution will have keys be tuples with 26 integers rather than however many characters. The details of how many bytes one integer takes up in Python is complicated, but from a quick investigation it's at least 24 bytes to account for Python's infinite precision (yeah it's a lot of overhead!). And that goes up as you need more to store the number, for example, I got 28 bytes for the number 13. So assuming 24 bytes, that's 26*24 for 624 bytes per key. (Other languages don't promise infinite precision this way, and integers are usually 4 bytes. That's still 104 bytes per key!)

However, characters always take either 1, 2, or 4 bytes. Though assuming English letters, it would be 1 byte. So assuming about 5 letters per word, that's 5 bytes per key (20 if you assume the worst case 4 bytes).

And yes, these would both still be O(n) in space, but your solution will take up 30 times the space by these rough numbers.

But, this is kind of just trivia, really, and I'm not taking off any points for this, I just wanted to explain a bit about how to think about size of your keys here. More important, the keys you're using are what are called "sparse arrays" in that the majority of values are going to be 0. Sometimes there's no choice, but if you have data that is mostly 0, it may be worth thinking about ways to save that data without saving all the zeros.